File: backend/src/ytforge/interfaces/api/schemas/assets.py
from __future__ import annotations

import uuid
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from ytforge.domain.enums import AssetStatus, AssetType


class AssetRegisterRequest(BaseModel):
    asset_type: AssetType
    bucket: str = Field(min_length=1, max_length=63)
    object_key: str = Field(min_length=1, max_length=1024)
    scene_id: uuid.UUID | None = None
    checksum_sha256: str | None = None
    provenance: dict[str, Any] = Field(default_factory=dict)


class AssetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    scene_id: uuid.UUID | None
    asset_type: AssetType
    status: AssetStatus
    bucket: str
    object_key: str
    checksum_sha256: str | None
    provenance: dict[str, Any]


class PresignedUrlRead(BaseModel):
    url: str


class ImageGenerateRequest(BaseModel):
    prompt: str = Field(min_length=1, max_length=2000)
    negative_prompt: str | None = None
    width: int = Field(default=1024, ge=64, le=2048)
    height: int = Field(default=1024, ge=64, le=2048)

File: frontend/src/lib/api/schemas/assets.ts
import { z } from "zod";
import { AssetStatus, AssetType } from "@/types/enums";

const assetTypeSchema = z.enum([
  AssetType.IMAGE,
  AssetType.CLIP,
  AssetType.AUDIO,
  AssetType.MUSIC,
  AssetType.THUMBNAIL,
  AssetType.RENDER,
]);

const assetStatusSchema = z.enum([
  AssetStatus.PENDING,
  AssetStatus.READY,
  AssetStatus.FAILED,
  AssetStatus.ORPHANED,
]);

export const assetReadSchema = z.object({
  id: z.string().uuid(),
  project_id: z.string().uuid(),
  scene_id: z.string().uuid().nullable(),
  asset_type: assetTypeSchema,
  status: assetStatusSchema,
  bucket: z.string(),
  object_key: z.string(),
  checksum_sha256: z.string().nullable(),
  provenance: z.record(z.string(), z.unknown()),
});
export type AssetRead = z.infer<typeof assetReadSchema>;

export type AssetRegisterRequest = {
  asset_type: AssetType;
  bucket: string;
  object_key: string;
  scene_id?: string | null;
  checksum_sha256?: string | null;
  provenance?: Record<string, unknown>;
};

export const presignedUrlReadSchema = z.object({ url: z.string() });
export type PresignedUrlRead = z.infer<typeof presignedUrlReadSchema>;

export type ImageGenerateRequest = {
  prompt: string;
  negative_prompt?: string | null;
  width?: number;
  height?: number;
};
